Firestore/core/src/firebase/firestore/api/input_validation.h

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_INPUT_VALIDATION_H_
```

Comment:
I was thinking a name like `error_reporting` might reflect what this file does better. However, I don't oppose the current name strongly, and I guess the fact that it matches TypeScript is a bonus.

Reply:
We have three classes of "error reporting":

  * runtime errors, which should be reported with `util::Status`
  * internal logic errors, which should be reported with `HARD_ASSERT`
  * user programming errors, which should be reported like this file does.

To me this means that `error_reporting` is ambiguous. I think it's important that this indicates that it's about only this last case. For example, internal argument checking should not `ThrowInvalidArgument`.